packages/database/src/migrations/20210125010541-ResyncDelayedSubmissionsAfterRelease70-modifies-data.js

```diff
@@ -0,0 +1,156 @@
+'use strict';
+
+import path from 'path';
+
+import { DatabaseChangeChannel } from '../DatabaseChangeChannel';
+import { arrayToDbString, markRecordsForResync } from '../utilities';
+
+var dbm;
+var type;
+var seed;
+
+/**
+ * We receive the dbmigrate dependency from dbmigrate initially.
+ * This enables us to not have to rely on NODE_PATH.
+ */
+exports.setup = function (options, seedLink) {
+  dbm = options.dbmigrate;
+  type = dbm.dataType;
+  seed = seedLink;
+};
+
+/**
+ * This migration attempts to resync survey responses and answers
+ * that were not added in the DHIS queue because of:
+ *
+ * 1. Bug #2160 that was in place between 2021-01-14T22:36Z - 2021-01-21T04.35Z
+ * 2. A server malfunction that blocked syncing between 2020-12-20 - 2020-12-22
+ *
+ * Here, we resync unsynced submissions by users who attempted to sync in the above periods
+ *
+ * @see https://github.com/beyondessential/tupaia-backlog/issues/885#issuecomment-767197853
+ * @see https://github.com/beyondessential/tupaia-backlog/issues/2143
+ * @see https://github.com/beyondessential/tupaia-backlog/issues/2160
+ */
+
+const DATE_RANGES = [
+  ['2020-12-20', '2020-12-23'],
+  ['2021-01-14', '2021-01-22'],
+];
+
+const selectEventBasedSurveys = async db => {
+  const { rows } = await db.runSql(`
+    SELECT s.code from survey s
+    JOIN survey_screen ss on ss.survey_id = s.id
+    JOIN survey_screen_component ssc on ssc.screen_id = ss.id
+    JOIN question q on q.id = ssc.question_id
+    WHERE
+      q.type = 'PrimaryEntity' AND
+      can_repeat is FALSE or ssc.config::text like '%case%'`);
+
+  return rows;
+};
+
+const createSelectResponseStatement = columns => {
+  const requestTimeCondition = DATE_RANGES.map(
+    ([start, end]) => `request_time BETWEEN '${start}' AND '${end}'`,
+  ).join(' OR ');
+
+  return `
+    SELECT sr.${columns} FROM survey_response sr
+    JOIN survey s ON sr.survey_id = s.id
+    JOIN data_source ds ON ds.id = s.data_source_id
+    JOIN entity e on e.id = sr.entity_id
+    LEFT JOIN dhis_sync_queue dsq ON sr.id = record_id
+    WHERE
+    dsq.id IS NULL AND
+    ds.service_type = 'dhis' AND
```

Comment:
I didn't think of this earlier, but perhaps we should check any that are using the default pathway to dhis rather than an explicit data_source?
```suggestion
    LEFT JOIN data_source ds ON ds.id = s.data_source_id
    JOIN entity e on e.id = sr.entity_id
    LEFT JOIN dhis_sync_queue dsq ON sr.id = record_id
    WHERE
    dsq.id IS NULL AND
    (ds.service_type = 'dhis' OR ds.service_type IS NULL) AND
```
Or do all surveys have a `data_source` entry?

Reply:
All surveys have a data_source entry